### ISSUE TITLE
Docs: Fix docs for "no-unnecessary-type-constraint"

### ISSUE THREAD

**Issue by iway1** [typescript-eslint/typescript-eslint#5609]
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

https://typescript-eslint.io/rules/no-unnecessary-type-constraint/ calls using `any` as a type constraint "redundant" even though it can provide great value to the developer (because it enables type inference).

IE, if you do this:

```ts
function aFunction<T extends any>(v: T){ return v }
const aValue = aFunction("s");
```

`aValue` will have type "s", which can makes code more typesafe and is very useful in some cases. Compare that to:
```ts
function aFunction(v: any){ return v; };
const aValue = aFunction("s");
```

Now `aValue` has type `any`, because we didn't extend any.

`any` is actually a particularly useful type to extend, because there are some functions that should work with anything, and where we would like to have the type inferred.

So, the docs are pretty misleading when it comes to type parameters passed to functions. Extending any enables type inference, which improves typesafety and makes code more robust. The rule itself may or may not be bad, either way the docs are definitely completely wrong (since they claim it's "redundant" and "unnecessary" even though it can make code more robust).

### Affected URL(s)

https://typescript-eslint.io/rules/no-unnecessary-type-constraint/

**Comment by bradzacher** [comment]
I think you've misunderstood the intent of the rule.
When the rule says that the type constraint is redundant - it doesn't mean that your entire generic is redundant - it means that the `extends any` is redundant.

I.e. instead of `<T extends any>`, do `<T>` - as for 99.999% of cases they're exactly the same thing.
[playground example](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAQwGLmvMAVAFsqAUQA8oBTMAEwGcBBMATwB5tEzSKaVGA+ACgBuALkTYAlAG9EAJzJQQ0pAMQBfAFAQE1KCgBqyADYgyARkQBeFOkiwEeAiXJU6jPgCJqbsQG41Aej9EIKCAPQB+NTVQGywrDFswFn5hUUkZOQUlVQ0tHWR9IzIAJgs4mIR3Tx9-QODEcMiAxDA4ckRcOAB3RChcMnT5RUQBQ2NEGGpEbRgDAx6GAAd+5EmAA2xVxD4YADoyHcQPLzYBCkROmF64EB1e-tX2Jy5kRlWgA)

**Comment by JoshuaKGoldberg** [comment]
cc @iway1 - closing as it's been >1 month without comment. Feel free to post back here if you have a response, we can always reopen. Cheers!

**Comment by iway1** [comment]
@bradzacher yeah you're right for some reason I thought that not including `extends any` would make the type parameter required? Guess I never tried it without that. Oof. 

Appreciate the clarification.
